consensus/service/src/api/client_api_service.rs

```diff
@@ -308,8 +367,21 @@ impl ConsensusClientApi for ClientApiService {
                     ConsensusGrpcError::NotServing.into()
                 }
             } else {
-                self.handle_proposed_tx(msg)
-                    .or_else(ConsensusGrpcError::into)
+                let result = self.handle_proposed_tx(msg);
+                // The block present below rate-limits suspicious behavior.
+                if let Err(_err) = &result {
+                    let mut tracker = self.tracked_sessions.lock().expect("Mutex poisoned");
```

Comment:
:wrench: I usually lean on the 3 strike rule:
1. you're writing code
2. It might be coincidence that the logic is duplicate
3. You're out, refactor.

This is the 3rd time this patter of `if get() else put() && get()` has been present. This should have method which does this for you. 
Unfortunately due to the mutex this will be a little bit more work as it will need to return a session guard that implements drop, to free up the mutex.